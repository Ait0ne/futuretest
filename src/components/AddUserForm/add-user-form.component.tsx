import React, { useReducer } from 'react';

import {FormContainer} from './add-user-form.styles';
import {AddUserProps} from '../AddUser/add-user.component';
import {Input, Button} from '../../styles/base-styles';
import {formReducer, initialFormState} from './add-user.reducer';




const AddUserForm: React.FC<AddUserProps> = ({addRow}) => {
    const [formState, dispatch] = useReducer(formReducer, initialFormState)

    const handleInputChange = (event: React.ChangeEvent<HTMLInputElement>) => {
        console.log(event.target.name)
        dispatch({type: event.target.name as 'id' | 'firstName'| 'lastName' | 'email' | 'phone', payload: event.target.value})
    }

    const handleSubmit = (event:React.FormEvent) => {
        event.preventDefault()
        if (!(formState.id===''||formState.firstName===''||formState.lastName===''||formState.email===''||formState.phone==='')) {
            addRow({
                ...formState,
                description: '',
                address: {
                    city: '',
                    state: '',
                    streetAddress: '',
                    zip: ''
                }
            })
            dispatch({type: 'CLEAR'})
        }

    }

    return(
        <FormContainer onSubmit={handleSubmit}>
            <Input 
            type='text'
            name='id'
            value={formState.id}
            placeholder='Id'
            onChange={handleInputChange}
            />
            <Input 
            type='text'
            name='firstName'
            placeholder='FirstName'
            value={formState.firstName}
            onChange={handleInputChange}
            />
            <Input 
            type='text'
            name='lastName'
            placeholder='LastName'
            value={formState.lastName}
            onChange={handleInputChange}
            />
            <Input 
            type='text'
            name='email'
            placeholder='Email'
            value={formState.email}
            onChange={handleInputChange}
            />
            <Input 
            type='text'
            name='phone'
            placeholder='Phone'
            value={formState.phone}
            onChange={handleInputChange}
            />
            <Button
            disabled={formState.id===''||formState.firstName===''||formState.lastName===''||formState.email===''||formState.phone===''}
            >
                Добавить в таблицу
            </Button>
        </FormContainer>
    )
}

export default AddUserForm;
