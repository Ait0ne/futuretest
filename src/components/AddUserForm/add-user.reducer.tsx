export interface FormState {
    id: string,
    firstName: string,
    lastName: string,
    phone: string,
    email: string
}


export type formReducerAction = {type:'id', payload:string}|
                                {type:'firstName', payload:string}|
                                {type:'lastName', payload:string}|
                                {type:'email', payload:string}|
                                {type:'phone', payload:string}|
                                {type:'CLEAR'}

export const initialFormState:FormState = {
    id: '',
    firstName: '',
    lastName: '',
    phone: '',
    email: ''
}

export const formReducer = (state:FormState, action:formReducerAction) => {
    console.log(action.type)
    switch(action.type) {
        case "id": 
            return {
                ...state,
                id: action.payload
            }
        case "firstName": 
            return {
                ...state,
                firstName: action.payload
            }
        case "lastName": 
            return {
                ...state,
                lastName: action.payload
            }
        case "phone": 
            return {
                ...state,
                phone: action.payload
            }
        case "email": 
            return {
                ...state,
                email: action.payload
            }
        case 'CLEAR': 
            return {
                ...initialFormState
            }
        default:
            return state
    }   
}