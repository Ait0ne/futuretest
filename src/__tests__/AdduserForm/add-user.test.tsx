import React from 'react';
import {mount} from 'enzyme';
import AddUserForm from '../../components/AddUserForm/add-user-form.component';
import {AddUserProps} from '../../components/AddUser/add-user.component';

const props:AddUserProps = {
    addRow:jest.fn()
}


describe('empty form', () => {
    const addUserForm = mount(<AddUserForm {...props}/>)

    
    it('renders correctly', () => {
        expect(addUserForm).toMatchSnapshot()
    })


    describe('inputs', () => {
        const inputs = addUserForm.find('input')
        
        it('renders 5 input fields', ()=> {
            expect(inputs.length).toEqual(5)
        })

        inputs.forEach((input)=> {
            it('is empty', () => {
                expect(input.prop('value')).toEqual('')
            })
        })

    })
    
    describe('button', () => {
        const button = addUserForm.find('button')
        
        it('renders a button', () => {
            expect(button.length).toEqual(1)
        })
        
        it('is disabled', () => {
            expect(button.prop('disabled')).toEqual(true)
        })
    })

    
    it('doesnt submit form', () => {
        addUserForm.simulate('submit')
        expect(props.addRow).toHaveBeenCalledTimes(0)
    })
})

describe('filled form', () => {
    const addUserForm = mount(<AddUserForm {...props}/>)
    const inputs = addUserForm.find('input')
    
    inputs.forEach((input:any) => {
        input.instance().value = 'hello'
        input.simulate('change')
    })

    
    
    describe('button', () => {
        const button = addUserForm.find('button')
        it('should have active button', ()=> {
            expect(button.prop('disabled')).toEqual(false)
        })

        it('should submit form with correct values', () => {
            button.simulate('submit')

            expect(props.addRow).toHaveBeenLastCalledWith({"address": {"city": "", "state": "", "streetAddress": "", "zip": 
            ""}, "description": "", "email": "hello", "firstName": "hello", "id": "hello", "lastName": "hello", "phone": "hello"})
        })
    })

    inputs.forEach((input:any)=> {
        it('should be empty', () => {
            expect(input.prop('value')).toEqual('')
        })
    })

})
