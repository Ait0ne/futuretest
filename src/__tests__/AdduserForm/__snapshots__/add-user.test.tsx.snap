// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`empty form renders correctly 1`] = `
<AddUserForm
  addRow={[MockFunction]}
>
  <styled.form
    onSubmit={[Function]}
  >
    <form
      className="sc-bdnylx dEMiuQ"
      onSubmit={[Function]}
    >
      <styled.input
        name="id"
        onChange={[Function]}
        placeholder="Id"
        type="text"
        value=""
      >
        <input
          className="sc-dlnjPT lovDsq"
          name="id"
          onChange={[Function]}
          placeholder="Id"
          type="text"
          value=""
        />
      </styled.input>
      <styled.input
        name="firstName"
        onChange={[Function]}
        placeholder="FirstName"
        type="text"
        value=""
      >
        <input
          className="sc-dlnjPT lovDsq"
          name="firstName"
          onChange={[Function]}
          placeholder="FirstName"
          type="text"
          value=""
        />
      </styled.input>
      <styled.input
        name="lastName"
        onChange={[Function]}
        placeholder="LastName"
        type="text"
        value=""
      >
        <input
          className="sc-dlnjPT lovDsq"
          name="lastName"
          onChange={[Function]}
          placeholder="LastName"
          type="text"
          value=""
        />
      </styled.input>
      <styled.input
        name="email"
        onChange={[Function]}
        placeholder="Email"
        type="text"
        value=""
      >
        <input
          className="sc-dlnjPT lovDsq"
          name="email"
          onChange={[Function]}
          placeholder="Email"
          type="text"
          value=""
        />
      </styled.input>
      <styled.input
        name="phone"
        onChange={[Function]}
        placeholder="Phone"
        type="text"
        value=""
      >
        <input
          className="sc-dlnjPT lovDsq"
          name="phone"
          onChange={[Function]}
          placeholder="Phone"
          type="text"
          value=""
        />
      </styled.input>
      <styled.button
        disabled={true}
      >
        <button
          className="sc-gtssRu eOovzA"
          disabled={true}
        >
          Добавить в таблицу
        </button>
      </styled.button>
    </form>
  </styled.form>
</AddUserForm>
`;
